### ISSUE TITLE
Fail to calculate gradients of gradients .

### ISSUE THREAD

**Issue by mapleam** [pytorch/pytorch#36567]
## 🐛 Bug

Network S, Network T.

We first use cross entropy loss between S and T to update T from S.

Meanwhile we retain the graph(retain_graph=True), then calculate loss L2 of updated S.

We aim to backward L2 to get gradients of S. However, gradients of S is None.

## To Reproduce

```python
import torch.nn.functional as F

input = torch.randn(1, 3, 224, 224).cuda()

weight_original = torch.randn(16, 3, 3, 3).cuda()
weight_original.requires_grad = True
weight_tea = torch.randn(16, 3, 3, 3).cuda()
weight_tea.requires_grad = True

out1_o = F.conv2d(input, weight_original)
out1_t = F.conv2d(input, weight_tea)
loss1 = (out1_o - out1_t).mean()
loss1.backward(retain_graph=True, create_graph=True)

weight_tea.grad.zero_()

weight_original = weight_original + weight_original.grad

out2 = F.conv2d(input, weight_original)
loss2 = out2.mean()
loss2.backward()
print(weight_tea.grad)
```

## Expected behavior

weight_tea.grad is Not None, but some real values.

## Environment

[pip3] numpy==1.16.2
[pip3] numpydoc==0.8.0
[pip3] pytorch-fft==0.15
[pip3] torch==0.4.1
[pip3] torchvision==0.2.1
[conda] numpy                     1.17.0                   pypi_0    pypi
[conda] torch                     1.1.0                    pypi_0    pypi
[conda] torchvision               0.2.1                    pypi_0    pypi

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by ezyang** [comment]
Usually in a situation like this, `loss2` is not actually connected to `weight_tea`. My guess is that is has to do with `weight_tea.grad.zero_()`. If you want to functionally compute the gradient of `weight_tea` without accumulating it into the `grad` member, I'd recommend just using `torch.autograd.grad` instead.

**Comment by mapleam** [comment]
> Usually in a situation like this, `loss2` is not actually connected to `weight_tea`. My guess is that is has to do with `weight_tea.grad.zero_()`. If you want to functionally compute the gradient of `weight_tea` without accumulating it into the `grad` member, I'd recommend just using `torch.autograd.grad` instead.

Hi, Thanks very much for you reply!

We aim to backward gradients through ```loss2``` to ```weight_tea```. Because in our opnion, ```loss2``` is calculated by ```weight_original + weight_original.grad``` and ```weight_original.grad``` is calculated by ```weight_tea```.

Then we tried codes below using only ```torch.autograd.grad``. However, ``weight_tea.grad``` is still None.
```
import torch
import torch.nn.functional as F

input = torch.randn(1, 3, 224, 224).cuda()

weight_original = torch.randn(16, 3, 3, 3).cuda()
weight_original.requires_grad = True
weight_tea = torch.randn(16, 3, 3, 3).cuda()
weight_tea.requires_grad = True
out1_t = F.conv2d(input, weight_tea)
out1_o = F.conv2d(input, weight_original)
loss1 = (out1_t - out1_o).mean()
# loss1.backward(retain_graph=True, create_graph=True)
gradd = torch.autograd.grad(loss1, weight_original)

weight_original_2 = weight_original + gradd[0]
out2 = F.conv2d(input.clone(), weight_original_2)
loss2 = out2.mean()
loss2.backward()
print(weight_tea.grad)
```

**Comment by albanD** [comment]
Hi,

We use github issues only for bugs or feature requests.
Please use the forum to ask questions: https://discuss.pytorch.org/
We will be happy to answer to you on the forum if you have follow up questions.
